### PR TITLE
PoPCHA HTTP auth server with in-memory storage

### PR DESCRIPTION
**Pull request by Nazianzenov**
- uncomplete storage
- no communication with the app yet
- unit testing
- property-based testing

### CODE REVIEW

File: be1-go/popcha/storage/mod.go [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+	popTokenToIdentifier map[string]string //hash map enforcing uniqueness on (PopToken,identifier)
+	nonces               map[string]struct{}
+	log                  zerolog.Logger
+}
+
+// NewStorage initializes a Storage object given a logger
+func NewStorage(log zerolog.Logger) (*Storage, error) {
+	return &Storage{
+		lock:                 &sync.Mutex{},
+		popTokenToIdentifier: make(map[string]string, 0),
```
Comment (MariemBaccari) [inline be1-go/popcha/storage/mod.go:25]: Is there a reason why you specify the 0 argument in `make()` for some usages and others not ? I think 0 is already a default value but I'm not 100% sure about that.

**Review by MariemBaccari — CHANGES_REQUESTED** [review]
The code looks generally clean and well structured. The comments help users who don't have experience with the POPCHA subproject understand the big picture of what's happening. I made a few comments about some small things that can be fixed before I approve the PR.

File: be1-go/popcha/server.go [hunk 2]
```diff
@@ -0,0 +1,410 @@
...
+	"sync"
+	"time"
+)
+
+const (
+	// formatting for auth request validation error
+	errValidAuthFormat = "Error while validating the auth request: %s"
+
+	//error message for invalid response type
+	errInvalidRepsonseType = "response type is invalid for the implicit flow"
```
Comment (MariemBaccari) [inline be1-go/popcha/server.go:31]: I already made this comment in my first review but there's this small typo here :)

File: be1-go/popcha/server.go [hunk 3]
```diff
@@ -0,0 +1,410 @@
...
+		as.handleBadRequest(w, err, "Error while generating PoPCHA QRCode")
+	}
+}
+
+func (as *AuthorizationServer) handleBadRequest(w http.ResponseWriter, err error, msg string) {
+	as.log.Err(err).Msg(msg)
+	w.WriteHeader(badRequestCode)
+	_, err = w.Write([]byte(strconv.Itoa(badRequestCode) + " - " + err.Error()))
+	if err != nil {
+		as.log.Err(err).Msg("Error while writing error message in the HTTP repsonse body")
```
Comment (MariemBaccari) [inline be1-go/popcha/server.go:274]: Same "repsonse" typo here as well

File: be1-go/popcha/server.go [hunk 4]
```diff
@@ -0,0 +1,410 @@
...
+	ResponseType = "response_type"
+	State        = "state"
+	ResTypeMulti = "id_token token"
+	OpenID       = "openid"
+	Profile      = "profile"
+	LoginHint    = "login_hint"
+)
+
+// clientParams implements op.Client
+// as PoPCHA don't support registered clients, clientParams is an alternative which defines  parameters
```
Comment (MariemBaccari) [inline be1-go/popcha/server.go:63]: I might be nitpicking here but there are two spaces between "defines" and "parameters"

File: be1-go/popcha/server.go [hunk 5]
```diff
@@ -0,0 +1,410 @@
...
+// for any client sending an auth request. As an example, any client requesting authorization has to use
+// the implicit flow.
+type clientParams struct {
+	clientID     string
+	redirectURIs []string
+	resType      oidc.ResponseType
+}
+
+/*
+Most of the methods of op.Client are implemented, however some of them
```
Comment (MariemBaccari) [inline be1-go/popcha/server.go:73]: Small suggestion: I think that physically separating the functions that are implemented and the ones that are not by some comments (in two different "blocs") would make the code more readable. Future programmers might find it useful to only see the functions they can directly use and the ones that just panic separately.

**Review by MariemBaccari — CHANGES_REQUESTED** [review]
Nice job ! Thanks for adding the comments. I wrote a few comments about typos and readability but the rest looks good to me.

File: be1-go/popcha/server.go (hunk 2 above)
Comment (Nazianzenov) [reply be1-go/popcha/server.go:31]: Oh my bad, i will correct that asap

File: be1-go/popcha/server.go (hunk 3 above)
Comment (Nazianzenov) [reply be1-go/popcha/server.go:274]: same, i will correct that!

File: be1-go/popcha/server.go (hunk 4 above)
Comment (Nazianzenov) [reply be1-go/popcha/server.go:63]: i think i use Goland IDE formatting, i will take a look at that !

File: be1-go/popcha/server.go (hunk 5 above)
Comment (Nazianzenov) [reply be1-go/popcha/server.go:73]: True! That's a very good suggestion definitely.

**Review by MariemBaccari — DISMISSED** [review]
Looks good to me now !

**Review by Anghul — CHANGES_REQUESTED** [review]
Good job overall. Good, well commented tests. Just a few comments before being able to approve it.

File: be1-go/popcha/server.go [hunk 6]
```diff
@@ -0,0 +1,416 @@
...
+	"sync"
+	"time"
+)
+
+const (
+	// formatting for auth request validation error
+	errValidAuthFormat = "Error while validating the auth request: %s"
+
+	//error message for invalid response type
+	errInvalidResponseType = "response type is invalid for the implicit flow"
```
[the remark this answers is not in the document]
Comment (Nazianzenov) [reply be1-go/popcha/server.go:31]: Unfortunately, the library doesn't cover the specific check for the response type we use. However, I can change the string and specify which response type is valid.

File: be1-go/popcha/server.go [hunk 7]
```diff
@@ -0,0 +1,416 @@
...
+	ClientID     = "client_id"
+	Nonce        = "nonce"
+	Scope        = "scope"
+	RedirectURI  = "redirect_uri"
+	ResponseType = "response_type"
+	State        = "state"
+	ResTypeMulti = "id_token token"
+	OpenID       = "openid"
+	Profile      = "profile"
+	LoginHint    = "login_hint"
```
[the remark this answers is not in the document]
Comment (Nazianzenov) [reply be1-go/popcha/server.go:59]: It is very likely i will need those outside, in order to avoid duplicating those variable.

**Review by Anghul — APPROVED** [review]
LGTM now

**Pull request merged**